### ISSUE TITLE
The range of session 

### ISSUE THREAD

**Issue by hbahn** [erlio/vernemq#155]
Hi there,

I read the definition of session in MQTT from VerneMQ homepage. (https://vernemq.com/intro/mqtt-primer/clean_session.html) Especially session information that stored in client.

However, I realized that this definition mentioned above is a bit different what MQTT speicification has specified regarding cleanSession.

The spec says

```
If not set (0), then the server must store the subscriptions of the client after it disconnects. This includes continuing to store QoS 1 and QoS 2 messages for the subscribed topics so that they can be delivered when the client reconnects. The server must also maintain the state of in-flight messages being delivered at the point the connection is lost. This information must be kept until the client reconnects.
```

Based on our test verneMQ does not store the subscription information even I set the persistent set in vernemq.conf.

is that right? and is there any plan to make the broker store all subscription when clean session is set to false in client side when connect.

**Comment by ioolkos** [comment]
Hi :)

let's layout things here a little: 

1. if a client connects with clean_session = false, then VerneMQ will have the correspondent topic subscriptions for the client (if there are any)

2. at this point, the server might have messages for the client, or not. We call these offline messages. The client will get all the offline messages upon re-connect (well, if it wants them & connects accordingly)

3. 'persistent_client_expiration' lets you define a time period, after which the broker should forget about an offline client. The default is 'never'.

I suspect what you're testing is case 2 (getting offline messages after re-connect). How do you test that?

Also don't forget that to store offline messages & later deliver them, you need to have this 3 conditions:

1. you need to connect with clean session false.
2. you need to subscribe with QoS > 0 and
3. you need to publish with QoS > 0

**Comment by hbahn** [comment]
Thanks for your explanation.

Could you check whether my scenario is correct or not?

1) a clinet connects with the option cleanSession = false and subscribes a few topics (QoS 1) and publish some messages.
2) the client disconnects.
3) When the client reconnects with the same client ID, is the previous session

**Comment by ioolkos** [comment]
re 3): if the session state should be resumed? Yes.

Let us know, if you experience a problem.

**Comment by hbahn** [comment]
Sorry I just pushed the wrong button when I was typing on my smartphone... :)

**Comment by ioolkos** [comment]
Apologies, missed your edits!

RE: "3) AFAIK, when the client reconnects with the same client ID, the previous session will be restored and In this particular situation, the client does not need to re-subscribe the topics after the re-connection since VerneMQ stores the topic subscriptions. is it right?"

Yes, this is correct. Note that if you test this with MQTT client tools, they might do application level things, like only subscribing explicitly, so they might give you a wrong idea.

You can verify with mosquitto_sub, subscribing to a topic, then disconnecting, then subscribing to a different topic. The subscriber will then still get messages published to the first topic.

**Comment by hbahn** [comment]
Could you share some solution to solve the problem for Clustering Situation?

In the normal case Clustering works well, (illustrated below)
![image003](https://cloud.githubusercontent.com/assets/13863157/16205979/48b75e8a-3761-11e6-8a8c-8e7dad2926e4.png)

Here is the problem (illustrated below) case that
![image006](https://cloud.githubusercontent.com/assets/13863157/16205998/64a0dec8-3761-11e6-8458-325fb5aba573.png)


1)     the client connected with node 1 in the cluster. Just disconnected due to device is failure. And the device just re-connects with node 2 in the cluster with client id B (no topic subscription again – clean session = false)

2)     All the publish message topic A is send to Node 1 and the newly re-connected device never gets any message.

3)     If the device re-connects with Node 1 again with client id B (no topic subscription again), the device suddenly receives the pending messages sent from devices (northbound) in the step2

Is there any solution for this case? We would like to see that even the device re-connects to a node in the cluster , the message is sent to the correct path not going to original subscription node…


Thanks in advance.
Hongbeom Ahn.

**Comment by ioolkos** [comment]
This should work exactly as you intend it to. Could you send the node's configuration files per mail to us?

**Comment by ioolkos** [comment]
I see what happens. You allow multiple sessions with the same client ID, but this in fact disables clean sessions 
https://github.com/erlio/vmq_server/blob/443250c6195db4e02dc1a05b561401da2737e628/src/vmq_reg.erl#L166

We'll have a discussion about this tomorrow.

**Comment by hbahn** [comment]
Thanks for your response.
We set these parameters,

1. trade_consistency = on
2. allows_multiple_session=on

as you understand a client connects with cleanSession=false when changing nodes in the cluster.

**Comment by ioolkos** [comment]
Yes. Let me explain this a little.

 `allow_multiple_sessions = on` disables the CleanSession feature if you connect with the same ClientID on a different cluster node.

Why is that?

Let's say you have 3 nodes A, B and C. Your client X is logged into node A and B. Now a connection comes in on node C, again with the same ClientID X. 
What session should it take over in case of Clean Session false, A or B? The answer is: the broker cannot decide.

So as a consequence the CleanSession feature is disabled. It still works on a per node basis though.

In other words: if you want to get messages for the same topics on a different node, you need to re-subscribe to them. 
This might seem unfortunate, and you may also have to deal with offline message delivery on other nodes. It's a consequence of trying to make the MQTT protocol (with its consistency requirements like single ClientID) 'distributable'. There's no easy way to do this.

As I said, we'll discuss this, but we will likely keep the semantics and behaviours like they are.

**Comment by ioolkos** [comment]
OK, we just had a discussion about the consistency/availability settings. We will make this semantically clearer by introducing a new option that separates allow_multiple_sessions_per_queue from an allow_registers_during_netsplits, or something similar. We can't do this short term though.

As the need you describe is basically the normal MQTT behaviour, we think simplest is to switch
allow_multiple_sessions off.

You're cluster will then still be semi-available during a network split (only disallowing new connections).

**Comment by hbahn** [comment]
I got you point. Thanks a lot!

Actually, our brokers(clustered) based on VerneMQ should connect with multiple Web-Application-Servers (distributed) with a clientID the broker publishes messages to the WASes in balanced manner (like HAproxy)

We hope to see that allow_multiple_sessions_per_queue from an allow_registers_during_netsplits near soon.
